fqdn: Improve performance by using selectors as labels

gandro opened the pull request:
This is an implementation of the following CFP. Please refer to the CFP for details on the approach.

- https://github.com/cilium/design-cfps/pull/17

It is highly recommended to consult the CFP, as it contains all the high-level design decisions and mechanism found in these commits. The commit messages themselves only consider low-level details. The main commit is rather large, as I was unable to split it into smaller chunks that would the test suite by themselves. However, it contains a commit message explaining each code chunk in the diff.

Follow-up work not intended to be part of this PR:

- [ ] ~Instead of collecting all matching selectors and every one as a label, extract the least generic available selector (e.g. `matchName: cilium.io` instead of `matchPattern: *`) for a given domain and use that one, as it should reduce identity churn. This requires us to teach the selector cache about selector hierarchy.~
- [ ] ~Pre-allocating an identity for a registered selector, rather than waiting for the first IP to be observed~ https://github.com/cilium/cilium/pull/33068#issuecomment-2173271245
- [x] Update to documentation regarding upgrade impact
- [x] Improve upgrade tests to also cover ToFQDN policies
- [x] Merge https://github.com/cilium/cilium/pull/32872 needed for dropless upgrades


```release-note
Improved performance for DNS lookups (up to 5x reduction in tail latency) when using ToFQDN policies. To avoid drops during upgrades in clusters with ToFQDN policies, it is highly recommended to run Cilium v1.15.6 or newer before upgrading to Cilium v1.16 
```

gandro committed:
labels: Simplify `IsReserved` implementation

This contians no functional changes and is a drive-by cleanup.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipcache: Always add world label to identities with fqdn label

A subsequent commit will change prefix labels upserted by the name
manager to use `fqdn`-labels instead of `cidr`-labels. Because a CIDR
identity currently always also have the world label, we want to mirror
that logic for identities with an `fqdn` label, as such IPs allowed
by a ToFQDN policy remains selectable by a `reserved:world` selector.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
fqdn: ToFQDN policy performance improvements

This commit implements `CFP-28427: ToFQDN policy performance
improvements`. It is highly recommended to consult the CFP, as it
contains all the high-level design decisions and mechanism found in this
commit. The rest of this commit message therefore only explains the
"what" and "where", and not the "why".

Before this commit, there was circular interaction between the
`SelectorCache` and `NameManager`: `SelectorCache` would tell
`NameManager` about new `ToFQDN` selectors, and `NameManager` would in
turn inform `SelectorCache` about the IPs selected by that `ToFQDN`
selector. This commit simplifies this logic by removing the backlink
from the `NameManager` to the `SelectorCache`. IPs are instead now
labelled with the selector as an `fqdn` identity label in IPCache, thus
not requiring any direct changes to the `SelectorCache` when a new IP is
discovered that shares the identity with an old IP. If there is identity
allocation needed for an observed IP, the `SelectorCache` is still
updated, but only via `IPCache`, and no longer directly from
`NameManager`.

I recommend first looking at the changes to `SelectorCache` in
`pkg/policy`. Note the following changes:

1. The `identityNotifier` interface (implemented by `NameManger`) is
   simplified: We no longer care about IPs selected by a FQDN selector,
   and we no longer need to care about potential deadlocks, as there are no
   calls back from `NameManager` to `SelectorCache` in the invoked
   functions (the indirect backlink from `NameManager` to `SelectorCache`
   via `IPCache` happens in `NameManager.UpdateGenerateDNS`
   - but this function is called by the DNS proxy whenever it observes a
    new DNS lookup and thus is called without the selector cache lock
   held.
2. `UpdateFQDNSelector` (previously invoked by `NameManager`) is removed
   - `SelectorCache` no longer directly needs to know the IPs matched by
   a selector.
3. The `fqdnSelector` type is simplified: Instead of containing the list
   of CIDR identities (one for each selected IP) and checking for the
   CIDR identity in `matches`, we now can simply treat the FQDN selector as
   a label and thus check if the requested identity has the FQDN selector
   label.
4. All the unit test logic around managing the selected IPs is removed,
   as all the responsibility for updating IPs now lies in `NameManager`.

For the `NameManager` in `pkg/fqdn`, the changes are as follows:

1.  Minor changes to for the query functions in `DNSCache`: Instead of
    just listing or checking the existence of an IP, we now want to know
    about `(name, IP)` pairs (needed later for updating `IPCache`).
2.  Similarly, where before we only cared about the mapping between an
    `FQDNSelector` and the selected IPs, we now want to know what
    `(name, IP)` pairs are matched by a particular selector.
    Thus `mapSelectorsToIPsLocked` is replaced with
    `mapSelectorsToNamesLocked` and the unit tests are updated as well.
3.  `RegisterFQDNSelector` now checks if the new selector needs to be
    added to any known `(name, IP)` pairs as an `fqdn` label, and
    `UnregisterFQDNSelector` potentially removes `fqdn` labels from IPs.
4.  `UpdateGenerateDNS` (invoked for DNS lookups) determines the labels
    of any newly discovered IP and now directly spawns the go routine
    to wait for the new `(IP, identity)` pair to be injected into
    `IPCache`. Previously, this waiting was done as part of the call to
    `UpdateSelectors`, previously implemented in `daemon/cmd/fqdn.go`
    (and now removed).
5.  `ForceGenerateDNS` is removed. It was previously called by the
    `NameManager` GC to remove IPs from the `SelectorCache`, but since
    the `SelectorCache` no longer knows about IPs, the function is
    obsolete (note that `IPCache` removals are still performed upon GC)
6.  Changes in `CompleteBootstrap` to deal with the upgrade logic
    when upgrading from Cilium v1.15. See bullet point 9 below for
    details.
7.  `updateDNSIPs` (called from `UpdateGenerateDNS`, i.e. upon new DNS
    lookups) now determines the labels for every newly observed IP based
    on the available FQDN selectors, and no longer upserts CIDR
    identites. Note that we only update the labels matching the looked up
    `dnsName`. If an IP happens to also map to a different domain name and
    uses a different set of selectors for the alternative name, those labels
    in IPCache are unaffected by the call to `updateMetadata`, as every call
    to IPCache uses the DNS name as the resource owner.
8.  The `ipcacheResource`, `updateMetadata`, and `maybeRemoveMetadata`
    contain the calls to `IPCache` to update labels for a given `(name,
    IP)` pair. There are two main differences to before: Instead of
    upserting or removing CIDR prefixes, we now add labels. And instead of
    having one update per prefix, we now have one update per `(name, IP)`
    pair, meaning a single prefix (aka "IP") might have multiple IPCache
    resource owners in the `NameManager` (i.e. one for each `name` mapping
    to that IP).
9.  `RestoreCache` and `CompleteBootstrap` contain the logic to
    initialize `IPCache` when upgrading from Cilium v1.15. This requires
    the previous Cilium instance to have checkpointed the known `ToFQDN`
    selectors, which are read in during upgrade and used to derive and inject the
    `IPCache` labels we expect to have once endpoint regeneration has
    finished. After endpoint regeneration, those restored labels are then
    removed, leaving the real labels in place.
    In contrast to all other `IPCache` updates (where each update to an
    IP is "owned" by the DNS name mapping to that IP, and we rely on
    `IPCache` to merge those labels), the resource owner here is static.
    This is, because they are all added at once (in `RestoreCache`)
    and removed at once (in `CompleteBootstrap`), and no per-name
    tracking is required.
10. Various changes to unit tests. The old unit tests tested the
    interaction between `NameManager` and `SelectorCache`, where as the
    new unit tests now test the interaction between `NameManager` and
    `IPCache`.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

squeed commented on daemon/cmd/fqdn.go:
```diff
@@ -54,64 +52,6 @@ const (
-//
-// returns a WaitGroup that is done when all policymaps have been updated
-// and all endpoints referencing the IPs are able to pass traffic.
-func (d *Daemon) updateSelectors(ctx context.Context, selectors map[policyApi.FQDNSelector][]netip.Addr, ipcacheRevision uint64) (wg *sync.WaitGroup) {
```
I am so glad to see this function gone :-D

squeed commented on pkg/fqdn/name_manager.go:
```diff
@@ -166,30 +165,15 @@ func (n *NameManager) GetDNSHistoryModel(endpointID string, prefixMatcher Prefix
+// associated with that selector.
+// This function also evaluates if any DNS names in the cache are matched by
+// this new selector and updates the labels for those DNS names accordingly.
+func (n *NameManager) RegisterFQDNSelector(selector api.FQDNSelector) {
```
We had discussed pre-allocating an identity for this selector. I think we should do this; it's a nice win and can help prevent policy drops on restart if IPs change identity.

gandro replied:
Oh yes, this fell between the cracks! Will add it to the PR

gandro replied:
Discussed offline, will add as a follow-up after this PR

gandro committed:
daemon: Wait for initial IPCache revision

This introduces a wait for the initial IPCache revision after K8s caches
have synced. This ensures that all prefix labels are injected and
available in the new IPCache before restoration starts.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
daemon: Also restore checkpointed FQDN identities

This commit modifies the IPCache restoration to restore all local
identity entries, not just CIDR identities.

This is required because FQDN labels are derived from ToFQDN selectors,
which are only available during endpoint regeneration. To ensure that
identities of prefixes in IPCache don't change during initial
regeneration, we provide the expected `fqdn` labels before regeneration.
The real labels are added during regeneration, therefore the restored
ones can be safely removed in `releaseRestoredIdentities`.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
fqdn: Derive domain labels from FQDN selectors

This commit adds logic to derive identity labels for `(name, IP)` pairs
from selectors. The basic idea is that any ToFQDN selector matching the
qname of the DNS lookup is added to a label to each IP returned by that
DNS lookup.

The functions added here will be used in a subsequent commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro commented on pkg/fqdn/name_manager.go:
```diff
@@ -262,49 +257,52 @@ func (n *NameManager) UpdateGenerateDNS(ctx context.Context, lookupTime time.Tim
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		n.config.IPCache.WaitForRevision(ipcacheRevision)
```
Like in the code before this PR, this go routine here is not cancelled if `ctx` expires. This is probably something we could change, though I might do that in a follow-up PR, which is why `ctx` has not been removed from this function signature (even though it us unused).

Note that the caller still aborts waiting on the waitgroup if `ctx` expires, so this is not a correctness issue (the only consequence is that we keep this go routine alive longer than we need to)

squeed replied:
If ctx is unused, we should remove it from the signature. There's currently no cancellable wait for a waitgroup anyways, so you have to churn a goroutine to two to do it. So just returning a waitgroup that the caller can use as they please is fine.

gandro replied:
My idea was to pass down `ctx` to `WaitForRevision` and then just call `Done` on the waitgroup even if `ctx` is cancelled. But this requires the caller then to check if if a timeout occurred once the `wg.Wait()` returns.

I think I might do that, it reduces the number of waiting go routines if the datapath is overloaded.

gandro replied:
Since this work (stopping the go routines early if datapath times out) is unrelated to the work of this PR (i.e. the problem exists on `main` too and is not affected in any way by this PR), I've started on it in a separate branch: https://github.com/cilium/cilium/compare/main...gandro:cilium:pr/gandro/dns-identities-followups?expand=1

Relevant commit: https://github.com/cilium/cilium/commit/bbd71e8537732bca5a8af668549d9e8ccba8561a

Marking this as resolved for now.

squeed commented on daemon/cmd/ipcache.go:
```diff
@@ -265,17 +265,19 @@ func (d *Daemon) restoreIPCache(localPrefixes map[netip.Prefix]identity.NumericI
-			if _, exists := id.Labels[wantLbl.Key]; exists {
+			_, isCIDRLabel := id.Labels[wantLbl.Key]
+			hasFQDNLabel := id.Labels.HasSource(labels.LabelSourceFQDN)
+			if isCIDRLabel || hasFQDNLabel {
```
Honestly, let's just remove this conditional; we can always restore these prefix + labels.

gandro replied:
Done!

squeed commented on pkg/fqdn/name_manager.go:
```diff
@@ -332,36 +328,37 @@ func (n *NameManager) updateDNSIPs(lookupTime time.Time, updatedDNSIPs map[strin
-	}
-
-	return affectedSelectors, updatedNames, ipcacheRevision
+	ipcacheRevision = n.updateMetadata(updatedMetadata)
```
Does it make sense to detect even more no-ops here? I realize this will short-circuit if the name's IPs haven't changed, but what about if the IP's labels haven't changed, just there's a new name.

This might be too much tracking, in which case it's probably not worth it. But it might be worth the bother. This can also be a follow-up PR.

gandro replied:
If there is a new name, then we ought to still update the IPCache metadata - since the new name needs to be a resource owner of the selector labels as well, even if it shares the same label selectors as the old name.

This is because the old name might be removed from the DNS cache[^1], in which case the prefix selector labels still need to be kept in the IPCache metadata. This is currently achieved by having each name be its own IPCache metadata resource owner. Thus, if there is a new name, then there is a new resource owner to keep the metadata alive.

We could of course add some tracking here to detect these cases here, but it would basically duplicate the IPCache metadata resource tracking. I'm not a fan of adding our custom tracking here if IPCache can do it for us. I'd rather have IPCache detect the no-op and skip the work (at least it skips the updates to BPF maps already).

[^1]: To be fair, I'm not entirely sure if the current DNS cache GC actually ever evicts a name if its IP is still actively used by another name. But at least from a lifecycle perspective I find it cleaner to assume it can, especially if we ever want to improve that behavior.

squeed replied:
Fair enough, we can file this away as a possible improvement if necessary. An opportunity to trade memory for CPU.

learnitall commented on pkg/fqdn/helpers.go:
```diff
@@ -8,83 +8,64 @@ import (
-					}
+		for dnsName, ips := range lookupIPs {
+			if len(ips) > 0 {
+				if log.Logger.IsLevelEnabled(logrus.DebugLevel) {
+					log.WithFields(logrus.Fields{
+						"DNSName":      dnsName,
+						"IPs":          ips,
+						"matchPattern": fqdnSelector.MatchPattern,
+					}).Debug("Emitting matching DNS Name -> IPs for FQDNSelector")
```
Thank you for wrapping this in a level check :heart_eyes:

learnitall commented on pkg/fqdn/name_manager_gc.go:
```diff
@@ -246,8 +264,85 @@ func (n *NameManager) RestoreCache(preCachePath string, restoredEPs []EndpointDN
+		if err != nil {
+			log.WithError(err).WithField(logfields.Path, checkpointPath).Error("Failed to restore FQDN selectors. " +
+				"Expect brief traffic disruptions for ToFQDN destinations during initial endpoint regeneration")
+			return
```
Would it make sense to add an option here to bail upgrading in favor of ensuring there are no traffic disruptions? I'm picturing that if a user wants to ensure there are no traffic disruptions when upgrading, this error log isn't going to be very helpful.

gandro replied:
Fair point. Let me see if we can add that to the pre-flight-check as a follow-up item. I think if we catch this before the user does an upgrade, this would be the best UX.

A flag on the agent itself will probably not work out, because bailing out on the upgrade is hard. Some agents might succeed and others won't. The ones that have succeeded we do not want to downgrade again, since downgrades are not dropless (unless we manage to merge https://github.com/cilium/cilium/pull/32914 and backport it), and then we're in a catch 22 where there is no path forward without causing drops on some nodes (as we currently don't have the ability to only upgrade some nodes that easily).

gandro replied:
Marking as resolved for now, as I prefer to do that check in a separate PR when we do the upgrade docs.

learnitall reviewed (approved):
This is fantastic, I'm so excited to see this merged. Question related to follow-ups: what kind of integration test coverage is provided for the 1.15 -> 1.16 upgrade path? I don't think the ci-e2e-upgrade workflow would provide coverage since we perform an upgrade without any policies applied to the cluster.

nebril reviewed (approved):
Simply awesome! ci-structure LGTM, everything else LGTM, :shipit:

squeed commented on pkg/fqdn/name_manager.go:
```diff
@@ -313,10 +311,9 @@ func (n *NameManager) CompleteBootstrap() {
 // corresponding set of IPs has changed.
```
nit: comment is wrong (affectedSelectors)

gandro replied:
Thanks, fixed!
